balloon/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+            self.hash_with_memory(pwd, salt, &mut memory)
+        } else {
+            #[cfg(not(feature = "parallel"))]
+            return Err(Error::ThreadsTooMany);
+            #[cfg(feature = "parallel")]
+            {
+                use core::num::NonZeroU32;
+                use rayon::iter::{IntoParallelIterator, ParallelIterator};
+
+                let mut params = self.params;
+                params.p_cost = NonZeroU32::new(1).unwrap();
+
+                (1..=self.params.p_cost.get())
+                    .into_par_iter()
+                    .map_with((params, self.secret), |(params, secret), index| {
+                        const U32_SIZE: usize = core::mem::size_of::<u32>();
+
+                        // expand salt if necessary
+                        let mut salt = if salt.len() < U32_SIZE {
+                            let mut new_salt = [0; U32_SIZE];
+                            new_salt[..salt.len()].copy_from_slice(salt);
+                            new_salt.to_vec()
+                        } else {
+                            salt.to_vec()
+                        };
+
+                        // add thread index to salt
+                        let byte = salt[..U32_SIZE].try_into().unwrap();
+                        salt[..U32_SIZE].copy_from_slice(
+                            &(u32::from_le_bytes(byte).wrapping_add(index)).to_le_bytes(),
+                        );
+                        Self::new(*params, *secret).hash(pwd, &salt)
+                    })
+                    .try_reduce(GenericArray::default, |a, b| {
+                        Ok(a.into_iter().zip(b).map(|(a, b)| a ^ b).collect())
+                    })
+            }
+        }
+    }
+
+    /// Hash a password and associated parameters.
+    ///
+    /// This method takes an explicit `memory_blocks` parameter which allows
+    /// the caller to provide the backing storage for the algorithm's state:
+    ///
+    /// - Users with the `alloc` feature enabled can use [`Balloon::hash`]
+    ///   to have it allocated for them.
+    /// - `no_std` users on "heapless" targets can use an array of the [`GenericArray`] type
+    ///   to stack allocate this buffer.
+    pub fn hash_with_memory(
+        &self,
+        pwd: &[u8],
+        salt: &[u8],
+        memory_blocks: &mut [GenericArray<u8, D::OutputSize>],
+    ) -> Result<GenericArray<u8, D::OutputSize>> {
+        if self.params.p_cost.get() > 1 {
+            return Err(Error::ThreadsTooMany);
+        }
+
+        let s_cost = usize::try_from(self.params.s_cost.get()).unwrap();
+        let t_cost = usize::try_from(self.params.t_cost.get()).unwrap();
+
+        let mut digest = D::new();
+
+        // This is a direct translation of the `Balloon` from <https://eprint.iacr.org/2016/027.pdf> chapter 3.1.
+        // int delta = 3 // Number of dependencies per block
+        const DELTA: u32 = 3;
```

Comment:
Some implementations also make this configurable, should we do this too? What name should I give it in the PHC string?

Reply:
I say fix delta. I'm not certain implementors will know what they're doing by tweaking it